homeassistant/components/stream/worker.py

```diff
@@ -11,104 +11,190 @@
-            packet.stream = self._stream_buffer.vstream
-            self._stream_buffer.output.mux(packet)
-        elif packet.stream == self._audio_stream:
-            packet.stream = self._stream_buffer.astream
-            self._stream_buffer.output.mux(packet)
-
-    def flush(self, duration):
+        if packet.stream == self._input_video_stream:
+            packet.stream = self._output_video_stream
+            self._av_output.mux(packet)
+        elif packet.stream == self._input_audio_stream:
+            packet.stream = self._output_audio_stream
+            self._av_output.mux(packet)
+
+    def check_flush_part(self, packet):
+        """Check for and mark a part segment boundary and record its duration."""
+        byte_position = self._memory_file.tell()
+        if self._segment.last_write_pos == byte_position:
+            return
+        if self._segment.last_write_pos == 0:  # The beginning of the first moof
+            self._segment.last_write_pos = self._segment.first_moof_loc = byte_position
+            self._segment.init = self._memory_file.getbuffer()[:byte_position].tobytes()
+            # Fetch the latest StreamOutputs, which may have changed since the
+            # worker started.
+            self._outputs = self._outputs_callback().values()
+            for stream_output in self._outputs:
+                stream_output.put(self._segment)
+        else:  # These are the ends of the part segments
+            self._segment.byterange_part_map[self._segment.last_write_pos] = len(
+                self._segment.parts
+            )
+            self._segment.parts.append(
+                Part(
+                    # The end of a part segment will actually have occurred before the last video packet
+                    duration=(self._last_packet_dts - self._part_start_dts)
+                    * packet.time_base,
+                    independent=self._part_is_independent,
+                    http_range_start=self._segment.last_write_pos
+                    - self._segment.first_moof_loc,
+                    data=self._memory_file.getbuffer()[
+                        self._segment.last_write_pos : byte_position
+                    ].tobytes(),
+                )
+            )
+            self._segment.last_write_pos = byte_position
+            for output in self._outputs:
+                output.part_put()  # Only really used for hls, but no harm otherwise
+            self._part_start_dts = self._last_packet_dts
+            self._part_is_independent = False
+
+    def flush(self, duration, packet):
         """Create a segment from the buffered packets and write to output."""
-        self._stream_buffer.output.close()
-        segment = Segment(
-            self._sequence, self._stream_buffer.segment, duration, self._stream_id
+        self._av_output.close()
+        self._segment.duration = duration
+        # Also flush the part segment (need to close the output above before this)
```

Comment:
With the way the part segmenting is detected, there will be one extra part here after the `_av_output` is closed. We flush this similarly to above.
After that we can clean up the file including an explicit `_memory_file.close()` which we couldn't do before as the BytesIO was still used in the HLS queue. This explicit closing may lead to quicker recycling of the memory.